### ISSUE TITLE
StreamJsonRpc.RemoteInvocationException: Unable to cast object of type 'FinalState' to type 'InProgressState'.

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#49448]
VS version: `Microsoft Visual Studio Enterprise 2019 Int Preview Version 16.9.0 Preview 2.0 [30709.18.main]`

Don't have repro steps - just saw bunch of gold bars pop up while typing:

![image](https://user-images.githubusercontent.com/10605811/99452054-ea5d6100-28d7-11eb-9680-49c0a2fdf21f.png)

```
StreamJsonRpc.RemoteInvocationException: Unable to cast object of type 'FinalState' to type 'InProgressState'.
   at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__134`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at Microsoft.CodeAnalysis.Remote.BrokeredServiceConnection`1.<TryInvokeAsync>d__17`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
RPC server exception:
System.InvalidCastException: Unable to cast object of type 'FinalState' to type 'InProgressState'.
      at Microsoft.CodeAnalysis.SolutionState.CompilationTracker.<GetOrBuildDeclarationCompilationAsync>d__22.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.CodeAnalysis.SolutionState.CompilationTracker.<GetMetadataOnlyImageReferenceAsync>d__35.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
      at Microsoft.CodeAnalysis.SolutionState.CompilationTracker.<GetMetadataReferenceAsync>d__33.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at Microsoft.CodeAnalysis.SolutionState.CompilationTracker.<FinalizeCompilationAsync>d__31.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.CodeAnalysis.SolutionState.CompilationTracker.<GetOrBuildCompilationInfoAsync>d__23.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.CodeAnalysis.SolutionState.CompilationTracker.<GetCompilationSlowAsync>d__20.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.CodeAnalysis.Document.<GetSemanticModelAsync>d__23.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.CodeAnalysis.DocumentHighlighting.AbstractDocumentHighlightsService.<GetDocumentHighlightsInCurrentProcessAsync>d__1.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.CodeAnalysis.DocumentHighlighting.AbstractDocumentHighlightsService.<GetDocumentHighlightsAsync>d__0.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.CodeAnalysis.Remote.RemoteDocumentHighlightsService.<>c__DisplayClass2_0.<<GetDocumentHighlightsAsync>b__0>d.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.CodeAnalysis.Remote.BrokeredServiceBase.<RunServiceImplAsync>d__12`1.MoveNext()

```

**Comment by CyrusNajmabadi** [comment]
Oh that's terrifying.  @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski   @jasonmalinowski

**Comment by jasonmalinowski** [comment]
Will be fixed by #49198.

**Comment by sbhenderson** [comment]
I recognize that this issue is closed, so I am hoping that this is not the incorrect place to put this. However, I have recently started running into this same exact issue described by the original report with a different exception text. This is in VS 2019 Enterprise 16.9.4. This will sound superstitious, but the error seems to show up most often when opening files containing a partial class that has had code generation performed via C# source generators affecting them, again, in a partial way. I may be able to come up with an example repro, if needed.

>StreamJsonRpc.RemoteInvocationException: Object reference not set to an instance of an object.
>   at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__139`1.MoveNext()
>--- End of stack trace from previous location where exception was thrown ---
>   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>   at System.Threading.Tasks.ValueTask`1.get_Result()
>   at System.Runtime.CompilerServices.ConfiguredValueTaskAwaitable`1.ConfiguredValueTaskAwaiter.GetResult()
>   at Microsoft.CodeAnalysis.Remote.BrokeredServiceConnection`1.<TryInvokeAsync>d__17`1.MoveNext()
>--- End of stack trace from previous location where exception was thrown ---
>   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
>RPC server exception:
>System.NullReferenceException: Object reference not set to an instance of an object.
>      at Microsoft.CodeAnalysis.CodeLens.CodeLensReferencesService.<TryGetMethodDescriptorAsync>d__8.MoveNext()
>   --- End of stack trace from previous location where exception was thrown ---
>      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>      at Microsoft.CodeAnalysis.CodeLens.CodeLensReferencesService.<>c__DisplayClass9_0.<<FindReferenceMethodsAsync>b__0>d.MoveNext()
>   --- End of stack trace from previous location where exception was thrown ---
>      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>      at Microsoft.CodeAnalysis.CodeLens.CodeLensReferencesService.<FindAsync>d__1`1.MoveNext()
>   --- End of stack trace from previous location where exception was thrown ---
>      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>      at Microsoft.CodeAnalysis.Remote.RemoteCodeLensReferencesService.<>c__DisplayClass5_0.<<FindReferenceMethodsAsync>b__0>d.MoveNext()
>   --- End of stack trace from previous location where exception was thrown ---
>      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>      at System.Threading.Tasks.ValueTask`1.get_Result()
>      at Microsoft.CodeAnalysis.Remote.BrokeredServiceBase.<RunServiceImplAsync>d__12`1.MoveNext()
>   --- End of stack trace from previous location where exception was thrown ---
>      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>      at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
